game/debriefing.py

```diff
@@ -109,16 +109,43 @@ class StateData:
     base_capture_events: List[str]
 
     @classmethod
-    def from_json(cls, data: Dict[str, Any]) -> StateData:
+    def from_json(cls, data: Dict[str, Any], unit_map: UnitMap) -> StateData:
+        def clean_unit_list(unit_list: List[Any]) -> List[str]:
+            # Cleans list of units in state.json by
+            # - Removing duplicates. Airfields emit a new "dead" event every time a bomb
+            #   is dropped on them when they've already dead.
+            # - Normalise dead map objects (which are ints) to strings. The unit map
+            #   only stores strings
+            units = set()
+            for unit in unit_list:
+                units.add(str(unit))
+            return list(units)
+
+        killed_aircraft = []
+        killed_ground_units = []
+
+        # process killed units from S_EVENT_UNIT_LOST & S_EVENT_KILL
+        killed_units = clean_unit_list(data["unit_lost_events"] + data["kill_events"])
+        for unit in killed_units:  # organize killed units into aircraft vs ground
+            if unit_map.flight(unit) is not None:
+                killed_aircraft.append(unit)
+            else:
+                killed_ground_units.append(unit)
+
+        # process killed units from S_EVENT_CRASH
+        for unit in data["crash_events"]:  # no need to clean crash events
+            if unit not in killed_aircraft:
+                killed_aircraft.append(unit)
+
+        # process killed units from S_EVENT_DEAD
```

Comment:
Or this?

If there's a reason to listen to all four of `S_EVENT_UNIT_LOST`, `S_EVENT_KILL`, `S_EVENT_DEAD` and `S_EVENT_CRASH`, could you explain those in a comment here?

Reply:
The idea behind listening to all events is that it's not entirely clear what logic DCS uses to generate these events, and there's no guarantee that this logic won't change over time. To have a robust kill detection mechanism, the approach was to log all relevant events even though it results in duplicates. I've added a comment to this effect.